cpu/atmega_common/include/gpio_ll_arch.h

```diff
@@ -0,0 +1,275 @@
+     (((port) - ATMEGA_GPIO_BASE_A) / ATMEGA_GPIO_SIZE))
+#else
+#define GPIO_PORT(num) (ATMEGA_GPIO_BASE_A + (num) * ATMEGA_GPIO_SIZE)
+#define GPIO_PORT_NUM(port) (((port) - ATMEGA_GPIO_BASE_A) / ATMEGA_GPIO_SIZE)
+#endif
+
+static inline uword_t gpio_ll_read(gpio_port_t port)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    return p->pin;
+}
+
+static inline uword_t gpio_ll_read_output(gpio_port_t port)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    return p->port;
+}
+
+/**
+ * @brief   Check if a set/clear operation with the on the given port with the
+ *          given mask can be done in a single-cycle bit-set / bit-clear
+ *          instruction.
+ *
+ * @param   port    GPIO port to set/clear bits on
+ * @param   mask    Bitmask specifying the bits to set/clear
+ *
+ * The compiler can optimize a clear/set operation if all of the following is
+ * true:
+ *
+ * 1. The GPIO port number is known at compile time, so that it can be used as
+ *    immediate
+ * 2. The GPIO port is alphabetically smaller than H, otherwise its registers
+ *    are not withing the I/O space (and cannot fit the 5 bit immediate field)
+ * 3. The bitmask is known at compile time
+ * 4. Exactly one bit is set in the bitmask
+ */
+static inline bool _can_bitwise_access(gpio_port_t port, uword_t mask)
+{
+    if (IS_CT_CONSTANT(port)
+#ifdef ATMEGA_GPIO_BASE_H
+        && IS_CT_CONSTANT(port < ATMEGA_GPIO_BASE_H)
+        && (port < ATMEGA_GPIO_BASE_H) 
+#endif
+        && IS_CT_CONSTANT(mask)
+        && IS_CT_CONSTANT(__builtin_popcount(mask) == 1)) {
+        return __builtin_popcount(mask) == 1;
+    }
+
+    return 0;
+}
+
+static inline void gpio_ll_set(gpio_port_t port, uword_t mask)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    if (_can_bitwise_access(port, mask)) {
+            p->port |= mask;
+    }
+    else {
+        unsigned state = irq_disable();
+        p->port |= mask;
+        irq_restore(state);
+    }
+}
+
+static inline void gpio_ll_clear(gpio_port_t port, uword_t mask)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    if (_can_bitwise_access(port, mask)) {
+        p->port &= ~mask;
+    }
+    else {
+        unsigned state = irq_disable();
+        p->port &= ~mask;
+        irq_restore(state);
+    }
+}
+
+static inline void gpio_ll_toggle(gpio_port_t port, uword_t mask)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    p->pin = mask;
+}
+
+static inline void gpio_ll_write(gpio_port_t port, uword_t value)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    p->port = value;
+}
+
+static inline gpio_port_t gpio_get_port(gpio_t pin)
+{
+    return GPIO_PORT(pin >> 4);
+}
+
+static inline uint8_t gpio_get_pin_num(gpio_t pin)
+{
+    return pin & 0x0f;
+}
+
+static inline uword_t gpio_ll_prepare_write_all_outputs(gpio_port_t port,
+                                                        uword_t value)
+{
+    atmega_gpio_port_t *p = (void *)port;
+    unsigned state = irq_disable();
```

Comment:
Doesn't `gpio_ll_prepapre_*` require locking anyway? So why disable IRQs?

Reply:
Agreed, but I'll also update the wording in the thread safety section to reflect this.